### PR TITLE
[exporter/datadog] Add ECS Fargate source provider

### PR DESCRIPTION
**Pull request by mx-psi**
**Description:** 

- Refactor current logic to a concept of source that identifies where a telemetry payload comes from (host or serverless function)
- Adds ECS Fargate source provider on the preview hostname logic.

The source refactor is entirely internal. The ECS Fargate logic will be exposed by #10926 but this PR needs no changelog since this logic is behind a constant for now.

**Link to tracking Issue:** #10424

**Testing:** Added unit tests.

### CODE REVIEW

File: exporter/datadogexporter/internal/metadata/internal/ecs/provider.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		return true, nil
+	}
+
+	return false, fmt.Errorf("TMDE endpoint is queryable, but launch type is unavailable")
+}
+
+// Source returns the task ARN of the ECS Fargate task if on ECS Fargate.
+func (p *Provider) Source(ctx context.Context) (source.Source, error) {
+	if onECSFargate, err := p.OnECSFargate(ctx); !onECSFargate && err == nil {
+		// Empty hostname for ECS Fargate tasks
```
Comment (KSerrania) [inline exporter/datadogexporter/internal/metadata/internal/ecs/provider.go:64]: The comment doesn't seem to match the branch it's in?

File: exporter/datadogexporter/internal/model/translator/metrics_translator.go [hunk 2]
```diff
@@ -396,28 +405,25 @@ func (t *Translator) MapMetrics(ctx context.Context, md pmetric.Metrics, consume
...
 
-		if host != "" {
-			// Track hosts if the consumer is a HostConsumer.
+		var host string
+		switch src.Kind {
+		case source.HostnameKind:
+			host = src.Identifier
 			if c, ok := consumer.(HostConsumer); ok {
-				c.ConsumeHost(host)
+				c.ConsumeHost(src.Identifier)
```
Comment (KSerrania) [inline exporter/datadogexporter/internal/model/translator/metrics_translator.go:422]: Since we define `host` a few lines above, can we use it here?

**Review by KSerrania — APPROVED** [review]
LGTM

File: exporter/datadogexporter/internal/metadata/internal/ecs/provider.go (hunk 1 above)
Comment (mx-psi) [reply exporter/datadogexporter/internal/metadata/internal/ecs/provider.go]: Indeed, fixed

File: exporter/datadogexporter/internal/model/translator/metrics_translator.go (hunk 2 above)
Comment (mx-psi) [reply exporter/datadogexporter/internal/model/translator/metrics_translator.go]: Done!

**Review by codeboten — COMMENTED** [review]
Please resolve the conflicts 👍🏻

**Comment by mx-psi** [comment]
@codeboten done!

**Review by codeboten — APPROVED** [review]

**Pull request merged**